---
title: "Choufani"
author: "The methylation in ART meta-analysis group"
date: "6/10/2020"
output: 
  html_document:
    toc: true
theme: cosmo
---


# Introduction
In this report, I will take you through a re-analysis methylation data first described by 
[Choufani et al (2019)](https://academic.oup.com/hmg/article/28/3/372/5102486).

In their study, they analysed the DNA methylation patterns of 44 placenta samples conceived
naturally (NAT) and 44 placenta samples conceivced through artificial reproductive technologies (ART) (23 in vitro, 21 in vivo).

The platform used in the study is the Illumina Infinium HumanMethylation450k BeadChip assay.

The methylation data have been deposited to NCBI GEO repository accession number 
[GSE120250](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE120250).

# Loading packages
These packackes will help us to perform vital steps such as normalisation, filtering, 
differential analysis, etc, and provide information about the array probe annotaions.

```{r,packages}
suppressPackageStartupMessages({
    library("R.utils")
    library("missMethyl")
    library("limma")
    library("topconfects")
    library("minfi")
    library("IlluminaHumanMethylation450kmanifest")
    library("MethylToSNP")
    library("RColorBrewer")
    library("IlluminaHumanMethylation450kanno.ilmn12.hg19")
    library("eulerr")
    library("plyr")
    library("gplots")
    library("reshape2")
    library("beeswarm")
    library("GEOquery")
    library("readxl")
    
  })
# Annotation
ann450k <- getAnnotation(IlluminaHumanMethylation450kanno.ilmn12.hg19)
myann <- data.frame(ann450k[,c("UCSC_RefGene_Name","Regulatory_Feature_Group")])
promoters <- grep("Prom",myann$Regulatory_Feature_Group)
```


# Loading functions

These functions provide shortcuts to help with charts and other analysis. They will
eventually be shoved into another Rscript or package but can stay here for now.

```{r,load_functions}

# scree plot shows the amount of variation in a dataset that is accounted
# for by the first N principal components
myscree <- function(mx,n=10,main="") {
  pc <- princomp(mx)$sdev
  pcp <- pc/sum(pc)*100
  pcp <- pcp[1:10]
  barplot(pcp,cex.names = 1,las=2,ylim=c(0,60),
      ylab="percent (%) variance explained", main=main)
  text((0.5:length(pcp)*1.2),pcp,label=signif(pcp,3),pos=3,cex=0.8)
}


# Here is a function to make a volcano plot
make_volcano <- function(dm,name,mx) {
    sig <- subset(dm,adj.P.Val<0.05)
    N_SIG=nrow(sig)
    N_UP=nrow(subset(sig,logFC>0))
    N_DN=nrow(subset(sig,logFC<0))
    HEADER=paste(N_SIG,"@5%FDR,", N_UP, "up", N_DN, "dn")
    plot(dm$logFC,-log10(dm$P.Val),cex=0.5,pch=19,col="darkgray",
        main=name, xlab="log FC", ylab="-log10 pval")
    mtext(HEADER)
    grid()
    points(sig$logFC,-log10(sig$P.Val),cex=0.5,pch=19,col="red")
}

# Here is a function to make heatmaps 
make_heatmap <- function(dm,name,mx,n) {
  topgenes <-  rownames(head(dm[order(dm$P.Value),],n))
  ss <- mx[which(rownames(mx) %in% topgenes),]
  my_palette <- colorRampPalette(c("blue", "white", "red"))(n = 25)
  heatmap.2(ss,scale="row",margin=c(10, 10),cexRow=0.4,trace="none",cexCol=0.4,
      col=my_palette, main=name)
}

# make beeswarm charts
# dm = a limma differential meth object
# name = character name of the limma dm object
# mx = matrix of normalised data
# groups = a vector of factors corresponding to the cols in mx
# n = the number of top significant genes to plot (default = 15) 
make_beeswarms <- function(dm,name,mx,groups,n=15) {
    par(mar=c(3,3,1,1))
    NCOLS=5
    NROWS=floor(n/NCOLS)
    if (n %% NCOLS > 0) { NROWS <- NROWS + 1 }
    par(mfrow=c(NROWS, NCOLS))
    topgenes <-  rownames(head(dm[order(dm$P.Value),],n))
    ss <- mx[which(rownames(mx) %in% topgenes),]
    n <- 1:n
    g1name=levels(groups)[1]
    g2name=levels(groups)[2]
    g1dat <- ss[n,which(groups == g1name)]
    g2dat <- ss[n,which(groups == g2name)]
    g1l <-lapply(split(g1dat, row.names(g1dat)), unlist)
    g2l <-lapply(split(g2dat, row.names(g2dat)), unlist)

    for (i in n) {
      mydat <- list(g1l[[i]],g2l[[i]])
        beeswarm(mydat,ylim=c(0,1),cex=0.2, pch=19,
        las=2, cex.lab=0.6, main=names( g1l )[i] , 
        ylab="",labels = c(g1name,g2name))
      grid()
    }
}

# make beeswarm charts for best confects
# dm = a limma differential meth object
# name = character name of the limma dm object
# mx = matrix of normalised data
# groups = a vector of factors corresponding to the cols in mx
# n = the number of top significant genes to plot (default = 15) 
make_beeswarms_confects <- function(confects,name,mx,groups,n=15) {
    par(mar=c(3,3,1,1))
    NCOLS=5
    NROWS=floor(n/NCOLS)
    if (n %% NCOLS > 0) { NROWS <- NROWS + 1 }
    par(mfrow=c(NROWS, NCOLS))
    topgenes <-  head(confects$table,n)$name
    ss <- mx[which(rownames(mx) %in% topgenes),]
    n <- 1:n
    g1name=levels(groups)[1]
    g2name=levels(groups)[2]
    g1dat <- ss[n,which(groups == g1name)]
    g2dat <- ss[n,which(groups == g2name)]
    g1l <-lapply(split(g1dat, row.names(g1dat)), unlist)
    g2l <-lapply(split(g2dat, row.names(g2dat)), unlist)

    for (i in n) {
      mydat <- list(g1l[[i]],g2l[[i]])
        beeswarm(mydat,ylim=c(0,1),cex=0.2, pch=19,
        las=2, cex.lab=0.6, main=names( g1l )[i] , 
        ylab="",labels = c(g1name,g2name))
      grid()
    }
}

# this is a wrapper which creates three charts
# We will be adding more
make_dm_plots <- function(dm,name,mx,groups=groups,confects=confects) {
    make_volcano(dm,name,mx)
    make_beeswarms(dm ,name , mx , groups , n= 15)
    make_heatmap(dm , name , mx ,n = 50)
    make_beeswarms_confects(confects, name, mx, groups, n=15)
}  

# this is a function which will perform differential methylation analysis
# if you provide it with the right inputs
dm_analysis <- function(samplesheet,sex,groups,mx,name,myann,beta) {

    design <- model.matrix(~ sex + groups)
    mxs <- mx[,which( colnames(mx) %in% samplesheet$Basename )]
    fit.reduced <- lmFit(mxs,design)
    fit.reduced <- eBayes(fit.reduced)
    summary(decideTests(fit.reduced))
    dm <- topTable(fit.reduced,coef=3, number = Inf)
    dma <- merge(myann,dm,by=0)
    dma <- dma[order(dma$P.Value),]
    dm_up <- rownames(subset(dm,adj.P.Val<0.05 & logFC>0))
    dm_dn <- rownames(subset(dm,adj.P.Val<0.05 & logFC<0))
    length(dm_up)
    length(dm_dn)
    confects <- limma_confects(fit.reduced, coef=3, fdr=0.05)
    make_dm_plots(dm = dm ,name=name , mx=beta, groups= groups, confects=confects)
    dat <- list("dma"=dma, "dm_up"=dm_up, "dm_dn"=dm_dn, "confects"=confects)

}

```

# Data Import
Importing Data from GEO.
Raw IDAT files and accompanying Series Matrix publicly available on GEO.
```{r,data}
#Create Dirtectory
dir.create("GSE120250")
ARRAY_SAMPLESHEET="GSE120250/GSE120250_series_matrix.txt.gz"
# only download it if it is not present on the system
if ( !file.exists(ARRAY_SAMPLESHEET ) ) {
    DLFILE=paste(ARRAY_SAMPLESHEET,".gz",sep="")
    download.file("https://ftp.ncbi.nlm.nih.gov/geo/series/GSE120nnn/GSE120250/matrix/GSE120250_series_matrix.txt.gz",
        destfile = DLFILE)
    gunzip(DLFILE)
}

# Reading in GEO series matrix
gse <- getGEO(filename="/mnt/mnorris/ART_methylation/GSE120250/GSE120250_series_matrix.txt.gz")


ARRAY_DATA="GSE120250/GSE120250_RAW.tar"
# only download it if it is not present on the system
if ( !dir.exists("GSE120250/IDAT") ) {
  dir.create("GSE120250/IDAT")
  download.file("https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE120250&format=file",
    destfile = ARRAY_DATA)
    untar(exdir = "GSE120250/IDAT", tarfile = "GSE120250/GSE120250_RAW.tar",)
}

baseDir <- "./GSE120250"
targets <- pData(phenoData(gse))

mybase <- unique(gsub("_Red.idat.gz" ,"", gsub("_Grn.idat.gz", "" ,list.files("./GSE120250",pattern = "GSM",recursive = TRUE))))
targets$Basename <- paste("GSE120250/", mybase, sep = "")
targets <- subset(targets,`outlier:ch1`=="N")
head(targets$Basename)
rgSet <- read.metharray.exp(targets = targets)
```
```{r,norm,fig.cap="Figure 1. Normalisation of bead-array data with SWAN."}
mSet <- preprocessRaw(rgSet)
mSetSw <- SWAN(mSet,verbose=FALSE)
par(mfrow=c(1,2))
densityByProbeType(mSet[,1], main = "Raw")
densityByProbeType(mSetSw[,1], main = "SWAN")
```
## Filter probes

Here we are running parallel analyses, both including and excluding sex chromosomes.

```{r,filterprobes}
# include sex chromosomes
detP <- detectionP(rgSet)
keep <- rowSums(detP < 0.01) == ncol(rgSet)
mSetSw <- mSetSw[keep,]

# exclude SNP probes
#mSetSw <- mapToGenome(mSetSw)
#mSetSw_nosnp <- dropLociWithSnps(mSetSw)
#dim(mSetSw)
#dim(mSetSw_nosnp)
#mSetSw <- mSetSw_nosnp

# exclude sex chromosomes
keep <- !(featureNames(mSetSw) %in% ann450k$Name[ann450k$chr %in% c("chrX","chrY")])
mSetFlt <- mSetSw[keep,]
head(mSetFlt)
dim(mSetFlt)
targets
```

## Extracting Beta and M-values

```{r,beta_m_vals}
# include sex chromosomes
meth <- getMeth(mSetSw)
unmeth <- getUnmeth(mSetSw)
Mval <- log2((meth + 100)/(unmeth + 100))
beta <- getBeta(mSetSw)

# exclude sex chromosomes
meth <- getMeth(mSetFlt)
unmeth <- getUnmeth(mSetFlt)
Mval_flt <- log2((meth + 100)/(unmeth + 100))
beta_flt <- getBeta(mSetFlt)


```

## Remove probes that appear to overlap SNP

DOI: https://doi.org/10.1186/s13072-019-0321-6

```{r,snpfilt}
#testing
#test <- MethylToSNP(beta, gap.ratio = 0.5, gap.sum.ratio = 0.4, SNP=SNPs.147CommonSingle #,verbose=FALSE)
#nrow(test)
# include sex chromosomes
# evidence of SNP
#snp_probe_dat <- MethylToSNP(beta, gap.ratio = 0.5, gap.sum.ratio = 0.4,
#    SNP=SNPs.147CommonSingle ,verbose=FALSE)
#dim(snp_probe_dat)
#head(snp_probe_dat)
#snp_probes <- rownames(snp_probe_dat)
# >50% confidence of SNP
#nrow(snp_probe_dat[which(snp_probe_dat$confidence>0.5),])
# remove SNP probes
#Mval <- Mval[which(!rownames(Mval) %in% snp_probes),]
#dim(Mval)
# evidence of SNP
#snp_probe_dat <- MethylToSNP(beta_flt, SNP=SNPs.147CommonSingle ,verbose=FALSE)
#dim(snp_probe_dat)
#head(snp_probe_dat)
#snp_probes <- rownames(snp_probe_dat)
# >50% confidence of SNP
#nrow(snp_probe_dat[which(snp_probe_dat$confidence>0.5),])
# remove SNP probes
#Mval_flt <- Mval_flt[which(!rownames(Mval_flt) %in% snp_probes),]
#dim(Mval_flt)
```

## Removal of Outliers
```{r, outliers}

targets2 <- subset(targets,`outlier:ch1`=="N")
head(targets2)
head(Mval)
colnames(Mval) <- sapply(strsplit(colnames(Mval), "_"), "[[", 1)
colnames(Mval_flt) <- sapply(strsplit(colnames(Mval_flt), "_"), "[[", 1)
Mval <- Mval[,which(colnames(Mval) %in% rownames(targets2))]
Mval_flt <- Mval_flt[,which(colnames(Mval_flt) %in% rownames(targets2))]

```

# MDS analysis

[Multidimensional scaling(https://en.wikipedia.org/wiki/Multidimensional_scaling)  plot is a method used
to identify the major sources of variation in a dataset. In the MDS plots below, I will be plotting the 
first two dimensions (principal components [PCs]), with each sample label coloured either by ART 
classification, sex, ART and sex, and then array chip and then sample plate. 

We wil begin with MDS analysis including the sex chromosomes and then exclude them.

First, let's quantify the contribution of the major principal components. with a scree plot, we can see 
whether most of the variation is captured in the first two PCs or whether it is spread over more PCs.

## Scree Plot
```{r,scree,fig.cap="Figure 2. Screeplot shows contribution of top principal components to the overall variation in the experiment."}

par(mfrow=c(2,1))  
myscree(Mval,main="incl sex chr")
myscree(Mval_flt,main="excl sex chr")

```


## MDS by ART
```{r,mds1,fig.width = 8 ,fig.height = 8,fig.cap="Figure 3. MDS plot coloured by ART classification."}

sample_groups <- factor(targets2$characteristics_ch1.2)
head(sample_groups)                        
colour_palette=brewer.pal(n = length(levels(sample_groups)), name = "Paired")
colours <- colour_palette[as.integer(factor(targets2$art))]
plot(1,axes = FALSE,xlab="",ylab="",main="MDS by ART type")
legend("center",legend=levels(sample_groups),pch=16,cex=1.2,col=colour_palette)
mydist <- plotMDS(Mval, labels=rownames(targets2),col=colours,main="sex chromosomes included")
mydist_flt <- plotMDS(Mval_flt, labels=rownames(targets2),col=colours,main="sex chromosomes excluded")

```
## MDS by Sex
```{r,mds2,fig.width = 8 ,fig.height = 8, fig.cap="Figure 4. MDS plot coloured by sex."}
sample_groups <- factor(targets2$characteristics_ch1.1)
colour_palette=brewer.pal(n = length(levels(sample_groups)), name = "Paired")
colours <- colour_palette[as.integer(factor(targets2$characteristics_ch1.1))]
plot(1,axes = FALSE,xlab="",ylab="",main="MDS by sex")
legend("center",legend=levels(sample_groups),pch=16,cex=1.2,col=colour_palette)
plotMDS(mydist, labels=rownames(targets2),col=colours,main="sex chromosomes included")
plotMDS(mydist_flt, labels=rownames(targets2) ,col=colours,main="sex chromosomes excluded")
```

## MDS by ART and sex
```{r,mds3,fig.width = 8 ,fig.height = 8, fig.cap="Figure 5. MDS plot coloured by ART classification and sex."}
targets2$ARTgender <- paste(targets$`art treatment:ch1`,targets2$`gender:ch1`, sep=("_"))

targets2$ARTgender
sample_groups <- factor(targets2$ARTgender)
sample_groups

colour_palette=brewer.pal(n = length(levels(sample_groups)), name = "Paired")
colours <- colour_palette[as.integer(factor(targets2$ARTgender))]
plot(1,axes = FALSE,xlab="",ylab="",main="MDS by ART and sex")
legend("center",legend=levels(sample_groups),pch=16,cex=1.2,col=colour_palette)
plotMDS(mydist, labels=rownames(targets2), col=colours,main="sex chromosomes included")
plotMDS(mydist_flt, labels=rownames(targets2), col=colours,main="sex chromosomes excluded")

```


# Differential analysis

There are a few differential contrasts that would be of interest to us in this study:

* __NAT vs ART__: 

* __NAT vs in vitro__: .

* __NAT vs in vivo__: .

* __in vivo vs in vitro__: 

The differential analysis is centred around limma to identify differentially methylated probes. 
TopConfects was also run to obtain the probes with the largest confident effect (topconfect).
There are five outputs below:

1. Volcano plot (limma result).

2. Beeswarm plot (top probes by limma p-value).

3. Heatmap (top probes by limma p-value).

4. Beeswarm plot (top probes by topconfect ranking).

5. Heatmap (top probes by topconfect ranking). (TODO)

## Nat vs In vivo
```{r,nat_vs_in_vivo_incl,fig.width = 8 ,fig.height = 8, fig.cap="Figure 6: Differential analysis: Naturally conceived vs in vivo including sex chromosomes"}

# Include Sex Chromosomes
samplesheet <- subset(targets2,`art treatment:ch1`=="NA" | `art treatment:ch1`=="in vivo")
samplesheet$Basename <- sapply(strsplit(samplesheet$Basename, "/"), "[[", 3)
samplesheet$Basename <- sapply(strsplit(samplesheet$Basename, "_"), "[[", 1)

groups <- factor(samplesheet$`art treatment:ch1`,levels=c("NA","in vivo"))
sex <- factor(samplesheet$`gender:ch1`,levels=c("M","F"))
design <- model.matrix(~ sex + groups)
    mxs <- Mval[,which( colnames(Mval) %in% samplesheet$Basename )]
    fit.reduced <- lmFit(mxs,design)
    fit.reduced <- eBayes(fit.reduced)
    summary(decideTests(fit.reduced))
 
top_NA_vs_invivo_inc <- dm_analysis(samplesheet=samplesheet,
    sex=sex,groups=groups,mx=Mval,name="top_NA_vs_invivo_inc",
    myann=myann ,beta= beta)
head(top_NA_vs_invivo_inc$dma)
```

```{r,nat_vs_in_vivo_exc, fig.width = 8 , fig.height = 8, fig.cap="Figure 7: Differential analysis: Naturally conceived vs in vivo excluding sex chromosomes"}
# Exclude Sex Chromosomes
samplesheet <- subset(targets2,`art treatment:ch1`=="NA" | `art treatment:ch1`=="in vivo")
samplesheet$Basename <- sapply(strsplit(samplesheet$Basename, "/"), "[[", 3)
samplesheet$Basename <- sapply(strsplit(samplesheet$Basename, "_"), "[[", 1)

groups <- factor(samplesheet$`art treatment:ch1`,levels=c("NA","in vivo"))
sex <- factor(samplesheet$`gender:ch1`,levels=c("M","F"))
design <- model.matrix(~ sex + groups)
    mxs <- Mval_flt[,which( colnames(Mval) %in% samplesheet$Basename )]
    fit.reduced <- lmFit(mxs,design)
    fit.reduced <- eBayes(fit.reduced)
    summary(decideTests(fit.reduced))
top_NA_vs_invivo_exc <- dm_analysis(samplesheet=samplesheet,
    sex=sex,groups=groups,mx=Mval_flt,name="top_NA_vs_invivo_exc",
    myann=myann ,beta= beta)
head(top_NA_vs_invivo_exc$dma)
```

## Nat vs In vitro

```{r,nat_vs_in_vitro_exc, fig.width = 8 ,fig.height = 8, fig.cap="Figure 9: Differential analysis: Naturally conceived vs in vitro excluding sex chromosomes"}
# Exclude Sex Chromosomes
samplesheet <- subset(targets2,`art treatment:ch1`=="NA" | `art treatment:ch1`=="in vitro")
samplesheet$Basename <- sapply(strsplit(samplesheet$Basename, "/"), "[[", 3)
samplesheet$Basename <- sapply(strsplit(samplesheet$Basename, "_"), "[[", 1)

groups <- factor(samplesheet$`art treatment:ch1`,levels=c("NA","in vitro"))
sex <- factor(samplesheet$`gender:ch1`,levels=c("M","F"))
design <- model.matrix(~ sex + groups)
    mxs <- Mval_flt[,which( colnames(Mval_flt) %in% samplesheet$Basename )]
    fit.reduced <- lmFit(mxs,design)
    fit.reduced <- eBayes(fit.reduced)
    summary(decideTests(fit.reduced))
top_NA_vs_invitro_exc <- dm_analysis(samplesheet=samplesheet,
    sex=sex,groups=groups,mx=Mval_flt,name="top_NA_vs_invitro_exc",
    myann=myann ,beta= beta)
head(top_NA_vs_invitro_exc$dma)
```

## Nat vs ART

```{r,ART_vs_Control_exc, fig.width = 8 ,fig.height = 8, fig.cap="Figure 11: Differential analysis: Naturally conceived vs ART excluding sex chromosomes"}
# Exclude Sex Chromosomes
samplesheet <- targets2
samplesheet$title <- as.character(samplesheet$title)
samplesheet$title <- (sapply(strsplit(samplesheet$title, " "), "[[", 1))
samplesheet$title <- (sapply(strsplit(samplesheet$title, "_"), "[[", 1))
samplesheet$Basename <- sapply(strsplit(samplesheet$Basename, "/"), "[[", 3)
samplesheet$Basename <- sapply(strsplit(samplesheet$Basename, "_"), "[[", 1)

groups <- factor(samplesheet$title,levels=c("ART","Control"))
sex <- factor(samplesheet$`gender:ch1`,levels=c("M","F"))
design <- model.matrix(~ sex + groups)
    mxs <- Mval_flt[,which( colnames(Mval_flt) %in% samplesheet$Basename )]
    fit.reduced <- lmFit(mxs,design)
    fit.reduced <- eBayes(fit.reduced)
    summary(decideTests(fit.reduced))
top_ART_vs_Control_exc <- dm_analysis(samplesheet=samplesheet,
    sex=sex,groups=groups,mx=Mval_flt,name="top_ART_vs_Natural_exc",
    myann=myann ,beta= beta)
head(top_ART_vs_Control_exc$dma)
```

## in vitro vs in vivo


```{r,in_vivo_vs_in_vitro_exc, fig.width = 8 ,fig.height = 8, fig.cap="Figure 13: Differential analysis: in vitro vs in vivo excluding sex chromosomes"}
# Exclude Sex Chromosomes
samplesheet <- subset(targets2,`art treatment:ch1`=="in vivo" | `art treatment:ch1`=="in vitro")
samplesheet$Basename <- sapply(strsplit(samplesheet$Basename, "/"), "[[", 3)
samplesheet$Basename <- sapply(strsplit(samplesheet$Basename, "_"), "[[", 1)
groups <- factor(samplesheet$`art treatment:ch1`,levels=c("in vivo","in vitro"))
sex <- factor(samplesheet$`gender:ch1`,levels=c("M","F"))
design <- model.matrix(~ sex + groups)
    mxs <- Mval_flt[,which( colnames(Mval_flt) %in% samplesheet$Basename )]
    fit.reduced <- lmFit(mxs,design)
    fit.reduced <- eBayes(fit.reduced)
    summary(decideTests(fit.reduced))
top_invivo_vs_invitro_exc <- dm_analysis(samplesheet=samplesheet,
    sex=sex,groups=groups,mx=Mval_flt,name="top_invivo_vs_invitro_exc",
    myann=myann ,beta= beta)
head(top_invivo_vs_invitro_exc$dma)
```

# Venn diagrams of the differential methylated probes for each contrast
First, we look at the similarity of DMPs altered between in vivo and in vitro. 

```{r,venn1,fig.cap="Figure 14:"}
#v1 <- list("in vivo up" = top_NA_vs_invivo_exc$dm_up , 
#           "in vitro up" = top_NA_vs_invitro_exc$dm_up ,
#           "in vivo dn" = top_NA_vs_invivo_exc$dm_dn ,
#           "in vitro dn" = top_NA_vs_invitro_exc$dm_dn )
#plot(euler(v1, shape = "ellipse"), quantities = TRUE)
```





