---
title: "Estill 2016 reanalysis"
author: "The methylation in ART meta-analysis group"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
theme: cosmo
---

## Introduction

In this report, I will take you through a re-analysis methylation data first described by 
[Estill et al (2016)](https://www.ncbi.nlm.nih.gov/pubmed/27288894).

In their study, they analysed the DNA methylation patterns of 137 neonatal blood spots
conceived naturally (NAT), through through insemination (IUI), or through ICSI using fresh or
cryopreserved (frozen) embryo transfer (FH and FZ respectively).

The platform used in the study is the Illumina Infinium HumanMethylation450k BeadChip assay.
The authors used a pipeline based on ChAMP 
([Tian et al, 2017](https://www.ncbi.nlm.nih.gov/pubmed/28961746)), together with Adjacent
Site Clustering ([Sofer et al, 2013](https://www.ncbi.nlm.nih.gov/pubmed/23990415)).

The methylation data have been deposited to NCBI GEO repository accession number 
[GSE79257](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE79257).

The main conclusions from the original study were:

* The methylation profiles of assisted reproductive technology and IUI newborns were dramatically
different from those of naturally (in vivo) conceived newborns.

* Profiles of ICSI-frozen (FET) and IUI infants were strikingly similar, suggesting that
cryopreservation may temper some of the epigenetic aberrations induced by IVF or ICSI.

*  The DNA methylation changes associated with IVF/ICSI culture conditions and/or parental
infertility were detected at metastable epialleles, suggesting a lasting impact on a child's
epigenome.

* Both infertility and ICSI alter DNA methylation at specific genomic loci, an effect that 
is mitigated to some extent by FET (freezing).

The aim of this work is to;

1. develop the analytical pipelines required for efficient re-analysis of 450K array data,

2. to confirm that we are able to obtain differential methylation results that are similar
to those obtained in the original study, and

3. to critically evaluate the conclusions made in the original study.

In this report I will be using the missMethyl vignette as a guide to analyse this dataset 
([Phipson et al, 2015](https://www.ncbi.nlm.nih.gov/pubmed/26424855)).

## Loading packages

These packackes will help us to perform vital steps such as normalisation, filtering, 
differential analysis, etc, and provide information about the array probe annotaions.

```{r,packages}
suppressPackageStartupMessages({
    library("R.utils")
    library("missMethyl")
    library("limma")
    library("minfi")
    library("IlluminaHumanMethylation450kmanifest")
    library("RColorBrewer")
    library("IlluminaHumanMethylation450kanno.ilmn12.hg19")
})
# Annotation
ann450k <- getAnnotation(IlluminaHumanMethylation450kanno.ilmn12.hg19)
```

## Data import

TODO- add explanation

```{r,data}
ARRAY_SAMPLESHEET="GSE79257_Illumina_samplesheet.csv"
# only download it if it is not present on the system
if ( !file.exists(ARRAY_SAMPLESHEET ) ) {
    DLFILE=paste(ARRAY_SAMPLESHEET,".gz",sep="")
    download.file("https://ftp.ncbi.nlm.nih.gov/geo/series/GSE79nnn/GSE79257/suppl/GSE79257_Illumina_samplesheet.csv.gz",
        destfile = DLFILE)
    gunzip(DLFILE)
}

ARRAY_DATA="GSE79257_RAW.tar"
# only download it if it is not present on the system
if ( !dir.exists("IDAT") ) {
  dir.create("IDAT")
  download.file("https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE79257&format=file",
    destfile = ARRAY_DATA)
}

untar(exdir = "IDAT", tarfile = "GSE79257_RAW.tar",)

baseDir <- "."
targets <- read.metharray.sheet(baseDir)
targets

rgSet <- read.metharray.exp(targets = targets)
rgSet

```

## Normalisation

```{r,norm,fig.cap="Figure 1. Normalisation of bead-array data with SWAN."}
mSet <- preprocessRaw(rgSet)
mSetSw <- SWAN(mSet,verbose=TRUE)
par(mfrow=c(1,2), cex=1.25)
densityByProbeType(mSet[,1], main = "Raw")
densityByProbeType(mSetSw[,1], main = "SWAN")
```

## Filter probes

Here we are running parallel analyses, both including and excluding sex chromosomes.

```{r,filterprobes}
# include sex chromosomes
detP <- detectionP(rgSet)
keep <- rowSums(detP < 0.01) == ncol(rgSet)
mSetSw <- mSetSw[keep,]

# exclude sex chromosomes
ann450k$Name
head(ann450k)
featureNames(mSetSw)
keep <- !(featureNames(mSetSw) %in% ann450k$Name[ann450k$chr %in% 
                                                     c("chrX","chrY")])
mSetFlt <- mSetSw[keep,]
```

## Extracting Beta and M-values

```{r,beta_m_vals}
# include sex chromosomes
meth <- getMeth(mSetSw)
unmeth <- getUnmeth(mSetSw)
Mval <- log2((meth + 100)/(unmeth + 100))
beta <- getBeta(mSetSw)
dim(Mval)

# exclude sex chromosomes
meth <- getMeth(mSetFlt)
unmeth <- getUnmeth(mSetFlt)
Mval_flt <- log2((meth + 100)/(unmeth + 100))
beta_flt <- getBeta(mSetFlt)
dim(Mval_flt)

```

## MDS plot

[Multidimensional scaling(https://en.wikipedia.org/wiki/Multidimensional_scaling)  plot is a method used
to identify the major sources of variation in a dataset. In the MDS plots below, I will be plotting the 
first two dimensions (principal components [PCs]), with each sample label coloured either by ART 
classification, sex, ART and sex, and then array chip and then sample plate. 

We wil begin with MDS analysis including the sex chromosomes and then exclude them.

First, let's quantify the contribution of the major principal components. with a scree plot, we can see 
whether most of the variation is captured in the first two PCs or whether it is spread over more PCs.
As we can see in Figure 2, the main source of variation is what is shown in PC1, and a much lesser
extent on the other dimensions.

```{r,scree,fig.cap="Figure 2. Screeplot shows contribution of top principal components to the overall variation in the experiment."}
par(mfrow=c(2,1))

pc<-princomp(Mval)$sdev
pcp <- pc/sum(pc)*100
pcp <- pcp[1:10]
barplot(pcp,cex.names = 1,las=2,ylim=c(0,60),ylab="percent (%) variance explained",
    main="include sex chromosomes")
text((0.5:length(pcp)*1.2),pcp,label=signif(pcp,3),pos=3,cex=0.6)

pc<-princomp(Mval_flt)$sdev
pcp <- pc/sum(pc)*100
pcp <- pcp[1:10]
barplot(pcp,cex.names = 1,las=2,ylim=c(0,60),ylab="percent (%) variance explained",
    main="exclude sex chromosomes")
text((0.5:length(pcp)*1.2),pcp,label=signif(pcp,3),pos=3,cex=0.6)

```


Here is the MDS plot by ART classification (Figure 3). You can see that there are four clusters 
of samples when MDS is projected this way. Natural birth infants (NAT) are mostly seen in the 
two clusters at the bottom of the chart and the IVF frozen, IVF fresh and in intrauterine 
insemination (IUI) appear mostly as two clusters at the top of the chart. IUI seem to have a
few samples that appear as intermediate between the upper and lower clusters.

```{r,mds1,fig.width = 8 ,fig.height = 8,fig.cap="Figure 3. MDS plot coloured by ART classification."}
targets$sex <- factor(sapply(strsplit(targets$Sample_Group,"_"),"[[",2))
targets$art <- factor(sapply(strsplit(targets$Sample_Group,"_"),"[[",1))

sample_groups <- factor(targets$art)

colour_palette=brewer.pal(n = length(levels(sample_groups)), name = "Paired")

colours <- colour_palette[as.integer(factor(targets$art))]

plot(1,axes = FALSE,xlab="",ylab="",main="MDS by ART type")

legend("center",legend=levels(sample_groups),pch=16,cex=1.2,col=colour_palette)

mydist <- plotMDS(Mval, labels=targets$Sample_Name,col=colours)
```

Next, we created an MDS plot by sex (Figure 4). The female samples appear in the two cluster on the 
right hand side of the chart and the male sample on the left. This could be simply due to the
chromosomal makeup of the male and female samples or could also be due to autosomes. To check this,
the MDS plot needs to be repeated with autosomes excluded.

```{r,mds3,fig.width = 8 ,fig.height = 8, fig.cap="Figure 4. MDS plot coloured by sex."}
sample_groups <- factor(targets$sex)

colour_palette=brewer.pal(n = length(levels(sample_groups)), name = "Paired")

colours <- colour_palette[as.integer(factor(targets$sex))]

plot(1,axes = FALSE,xlab="",ylab="",main="MDS by sex")

legend("center",legend=levels(sample_groups),pch=16,cex=1.2,col=colour_palette)

plotMDS(mydist, labels=targets$Sample_Name,col=colours)
```

For completeness, we show the MDS by ART and sex (Figure 5). It confirms the trends seen above where the 
samples are split on the x axis by sex and on the y axis by ART classification.

```{r,mds2,fig.width = 8 ,fig.height = 8, fig.cap="Figure 5. MDS plot coloured by ART classification and sex."}
sample_groups <- factor(targets$Sample_Group)

colour_palette=brewer.pal(n = length(levels(sample_groups)), name = "Paired")

colours <- colour_palette[as.integer(factor(targets$Sample_Group))]

plot(1,axes = FALSE,xlab="",ylab="",main="MDS by ART and sex")

legend("center",legend=levels(sample_groups),pch=16,cex=1.2,col=colour_palette)

plotMDS(mydist, labels=targets$Sample_Name,col=colours)
```

To acertain whether technical factors like batch effects account for variance on the two main PCs, we
create an MDS by array chip (Figure 6). There appears to be no relationship between array chip number
and the top two PCs.

```{r,mds2,fig.width = 8 ,fig.height = 8, fig.cap="Figure 6. MDS plot coloured by array chip."}
sample_groups <- factor(targets$Array)

colour_palette=brewer.pal(n = length(levels(sample_groups)), name = "Paired")

colours <- colour_palette[as.integer(factor(targets$Array))]

plot(1,axes = FALSE,xlab="",ylab="",main="MDS by array chip")

legend("center",legend=levels(sample_groups),pch=16,cex=1.2,col=colour_palette)

plotMDS(mydist, labels=targets$Sample_Name,col=colours)
```

This was also performed for sample plate (Figure 7). It appears that the laboratory did not randomise
samples over plates 1 and 2. For example the samples on plate 2 appear to be mostly NAT. This does
not appear to be a major concern because NAT samples from plate 2 seem to cluster with NAT samples on
plate 1.

```{r,mds2,fig.width = 8 ,fig.height = 8, fig.cap="Figure 7. MDS plot coloured by sample plate."}
sample_groups <- factor(targets$Sample_Plate)

colour_palette=brewer.pal(n = length(levels(sample_groups)), name = "Paired")

colours <- colour_palette[as.integer(factor(targets$Sample_Plate))]

plot(1,axes = FALSE,xlab="",ylab="",main="MDS by sample plate")

legend("center",legend=levels(sample_groups),pch=16,cex=1.2,col=colour_palette)

plotMDS(mydist, labels=targets$Sample_Name,col=colours)
```

```{r,autosomes}

str(IlluminaHumanMethylation450kmanifest)

data(IlluminaHumanMethylation450kmanifest)
keep <- !(featureNames(mSetSqFlt) %in% ann450k$Name[ann450k$chr %in% 
                                                        c("chrX","chrY")])
table(keep)
mSetSqFlt <- mSetSqFlt[keep,]

```





```{r,mytest}

#legend("topleft",legend=levels(sample_groups),pch=16,cex=1.2,col=1:2)

```
## References

Estill MS, Bolnick JM, Waterland RA, Bolnick AD, Diamond MP, Krawetz SA.
Assisted reproductive technology alters deoxyribonucleic acid methylation
profiles in bloodspots of newborn infants. Fertil Steril. 2016 Sep
1;106(3):629-639.e10. doi: 10.1016/j.fertnstert.2016.05.006. Epub 2016 Jun 8.
PubMed PMID: 27288894.

Phipson B, Maksimovic J, Oshlack A. missMethyl: an R package for analyzing
data from Illumina's HumanMethylation450 platform. Bioinformatics. 2016 Jan
15;32(2):286-8. doi: 10.1093/bioinformatics/btv560. Epub 2015 Sep 30. PubMed
PMID: 26424855.

Sofer T, Schifano ED, Hoppin JA, Hou L, Baccarelli AA. A-clustering: a novel
method for the detection of co-regulated methylation regions, and regions
associated with exposure. Bioinformatics. 2013 Nov 15;29(22):2884-91. doi:
10.1093/bioinformatics/btt498. Epub 2013 Aug 29. PubMed PMID: 23990415; PubMed
Central PMCID: PMC3810849.

Tian Y, Morris TJ, Webster AP, Yang Z, Beck S, Feber A, Teschendorff AE.
ChAMP: updated methylation analysis pipeline for Illumina BeadChips.
Bioinformatics. 2017 Dec 15;33(24):3982-3984. doi: 10.1093/bioinformatics/btx513.
PubMed PMID: 28961746; PubMed Central PMCID: PMC5860089.
