---
title: "METAl meta-Analysis"
author: "The methylation in ART meta-analysis group"
date: "10/1/2020"
output:   
  html_document:
    toc: true
theme: cosmo
---

# Introduction
Here we run the possible meta analysis comparisons

Match anno to probes
DMRcate peak calling
pvalue histogram
volcano plot
manhattan plot with split axis (split by z score) plot hypo hyper separately
Effectsize Deltabeta
Forrest Plot

bumpHunter, DMRfinder, probelasso, mCSEA, DMRforPairs 


```{r, install packages}

source("meth_functions.R")
ann450k <- getAnnotation(IlluminaHumanMethylation450kanno.ilmn12.hg19)
myann <- data.frame(ann450k[,c("chr", "pos", "UCSC_RefGene_Name","Regulatory_Feature_Group","Islands_Name","Relation_to_Island")])
library(bumphunter)

```

MARKER = Row.names
DEFAULT = N (Neff=4/(1/Ncases+1/Nctrls))
ALLELE = unmeth meth
FREQ = AveExpr
EFFECT = CE
STDERR = SE
PVAL = P.Value



# Natural vs ART
Here we compare all data sets with by Natural vs ART, as all studies use different methods this is a general comparison.
Estill
Choufani
Litzky
Hajj
```{bash, Natural vs ART} 

metal NATvsART.txt

```

```{r, Natural vs ART, output}
nat_vs_art <- read.table("METAANALYSIS1.TBL", header = TRUE)
nat_vs_art <- nat_vs_art[order(nat_vs_art$P.value),]
nat_vs_art$FDR <- p.adjust(nat_vs_art$P.value)


nat_vs_art_a <- merge(nat_vs_art, myann, by.x = "MarkerName", by.y = 0)
nat_vs_art_a <- nat_vs_art_a[order(nat_vs_art_a$P.value),]
head(nat_vs_art_a, 20)


hist(-log10(nat_vs_art_a$FDR))
hist(-log10(nat_vs_art_a$P.value))
hist(nat_vs_art_a$FDR)
hist(nat_vs_art_a$P.value, main = "Histogram of P.values")

volcano_metal(nat_vs_art_a, "Nat vs ART")


nat_vs_art_a$chrnum <- as.integer(gsub("chr","", nat_vs_art_a$chr))


manhattan(nat_vs_art_a, chr="chrnum", bp="pos", p="P.value")
manhattan(nat_vs_art_a, chr="chrnum", bp="pos", p="FDR")


hyper <- subset(nat_vs_art_a, Zscore > 0)
hypo <- subset(nat_vs_art_a, Zscore < 0)

manhattan(hyper, chr="chrnum", bp="pos", p="P.value", )
manhattan(hypo, chr="chrnum", bp="pos", p="P.value")
```

```{r, DMR calling}
# DMR
nat_vs_art_a[order(nat_vs_art_a$chr),]
res <- makeGRangesFromDataFrame(nat_vs_art_a, keep.extra.columns = TRUE, ignore.strand = TRUE, seqnames.field = "chr", start.field = "pos", end.field = "pos")
res <- sortSeqlevels(res)
res <- sort(res)
head(res$P.value)
resdf <- as.data.frame(res)

temp <- as.vector(apply(resdf, 1, function(x){
  if (x[11] > 0.05){y<-0}
  else if (x[10]> 0){y<- 1}
  else if (x[10]< 0){y<- -1}
  y
}))



dim(temp)
str(temp)
head(temp)

rolling <- rollmean(temp,k=3, fill = NA)
resdf$temp <- temp
resdf$rolling <- rolling
resdf

as.data.frame(resdf)[unname(which(rolling == 1)),]

dim(as.data.frame(res))

which(rolling == 1)   

```


```{r, Hyper}
hyper <- as.data.frame(t(sapply(which(rolling == 1), function(x){
  w=x-1
  y=x+1
  dmrstart <- start(res)[w]
  dmrend <- start(res)[y]
  chr <- seqnames(res)[y]
  return(c(w, x, y, dmrstart, dmrend))
})))


hyper_chr <- sapply(which(rolling == 1), function(x){
   chr <- seqnames(res)[x]
   as.character(chr)
})

hyper$chr <- hyper_chr
colnames(hyper)<- c("w", "x", "y", "start", "end", "chr")
hyper

hyperGR <- makeGRangesFromDataFrame(hyper, keep.extra.columns = TRUE, ignore.strand = TRUE, seqnames.field = "chr", start.field = "start", end.field = "end")
hyperGR <- GenomicRanges::reduce(hyperGR)
hyperGR
hyper


apply(as.data.frame(hyperGR), 1, function(x){
 G <- GRanges(seqnames = x[1], ranges = IRanges(x[2]:x[3]))
subsetByOverlaps(Granges, G)
 })



subsetByOverlaps(Granges, hyperGR)


hypo <- sapply(which(rolling == -1), function(x){
  w=x-1
  y=x+1
  dmrstart <- start(Granges)[w]
  dmrend <- start(Granges)[y]
  chr <- seqnames(Granges)[y]
  return(c(w, x, y, chr, dmrstart, dmrend))
})

hypo



 
```

# Natural vs ICSI 
Estill
Choufani
Hajj

```{bash, Natural vs ICSI}

metal NATvsICSI.txt

```

```{r, Natural vs ICSI, output}
nat_vs_icsi <- read.table("METAANALYSIS1.TBL", header = TRUE)
nat_vs_icsi <- nat_vs_icsi[order(nat_vs_icsi$P.value),]
head(nat_vs_icsi, 20)
```


# Natural vs Subfertility
Two of the studies also listed subfertility as a parameter, here we compare the two studies.
Choufani
Litzky
```{bash, Natural vs SUB}

metal NATvsSUB.txt

```

```{r, Natural vs Sub, output}
nat_vs_sub <- read.table("METAANALYSIS1.TBL", header = TRUE)
nat_vs_sub <- nat_vs_sub[order(nat_vs_sub$P.value),]
head(nat_vs_sub, 20)
```
Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.


# Natural vs ART in Placenta
Two of the studies have samples taken from placenta both of which when analysed individually have few-no significant probes, here we compare the two studies.
Choufani
Litzky
```{bash, Natural vs ART in placenta}

metal NATvsARTplacenta.txt

```

```{r, Natural vs ART placenta, output} 
nat_vs_artp <- read.table("METAANALYSIS1.TBL", header = TRUE)
nat_vs_artp <- nat_vs_artp[order(nat_vs_artp$P.value),]
head(nat_vs_artp, 20)
```
Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
