---
title: "Novakovic et al 2019 reanalysis"
author: "The methylation in ART meta-analysis group"
date: "`r Sys.Date()`"
output:
  html_document:
  toc: true
theme: cosmo
---
```{r,packages}
suppressPackageStartupMessages({
    library("R.utils")
    library("missMethyl")
    library("limma")
    library("minfi")
    library("IlluminaHumanMethylationEPICanno.ilm10b2.hg19")
    library("RColorBrewer")
    library("IlluminaHumanMethylationEPICanno.ilm10b2.hg19")
    library("eulerr")
    library("plyr")
    library("gplots")
    library("reshape2")
    library("beeswarm")
  library("GEOquery")
library("Biobase")
  })
```
# Obtaining array annotations

```{r,annotation}
anno <- getAnnotation(IlluminaHumanMethylationEPICanno.ilm10b4.hg19)
myann <- data.frame(anno[,c("UCSC_RefGene_Name","Regulatory_Feature_Group")])
```


## Introduction
  
This report is the reanalysis of first analysed DNA methylation data by
[Novakovic et al (2019)] (https://doi.org/10.1038/s41467-019-11929-9)

In this study,  DNA methylation status was generated for 149 neonatal (84♀ 65♂) and 158 adult (87♀ 71♂) 
ART-conceived individuals and for 58 neonatal (37♀, 21♂) and 75 adult (51♀, 24♂) non-ART conceived individuals.


```{r,download}
ARRAY_DATA="GSE131433_RAW.tar"
if(!dir.exists("IDAT")){
  download.file("https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE131433&format=file",destfile=ARRAY_DATA) 
  untar(exdir = "IDAT", tarfile = "GSE131433_RAW.tar",)
}
  SERIES_MATRIX="GSE1311433_series_matrix.txt.gz"
download.file("https://ftp.ncbi.nlm.nih.gov/geo/series/GSE131nnn/GSE131433/matrix/GSE131433_series_matrix.txt.gz",destfile=SERIES_MATRIX)


if(!file.exists("GSE1311433_series_matrix.txt.gz")){
gse <- getGEO(filename="/mnt/mkasu/projects/ART_methylation/GSE1311433_series_matrix.txt.gz")
}
baseDir <- "."
sample_metadata <- pData(phenoData(gse))
targets <- sample_metadata
mybase <- unique(gsub("_Red.idat.gz" ,"",  gsub("_Grn.idat.gz", "" ,list.files(".",pattern = "GSM",recursive = TRUE))))
targets$Basename <- mybase
head(targets)
rgSet <- read.metharray.exp(targets = targets)
rgSet

```

```{r,norm,fig.cap="Figure 1. Normalisation of bead-array data with SWAN."}
mSet <- preprocessRaw(rgSet)
mSetSw <- SWAN(mSet,verbose=FALSE)
par(mfrow=c(1,2), cex=0.8)
densityByProbeType(mSet[,1], main = "Raw")
densityByProbeType(mSetSw[,1], main = "SWAN")
```


```{r,filterprobes}
# include sex chromosomes
detP <- detectionP(rgSet)
keep <- rowSums(detP < 0.01) == ncol(rgSet)
mSetSw <- mSetSw[keep,]
# exclude SNP probes
mSetSw <- mapToGenome(mSetSw)
mSetSw_nosnp <- dropLociWithSnps(mSetSw)
# exclude sex chromosomes
keep <- !(featureNames(mSetSw) %in% myann$Name[myann$chr %in% 
                                                     c("chrX","chrY")])
mSetFlt <- mSetSw[keep,]
```

```{r,beta_m_vals}
# include sex chromosomes
meth <- getMeth(mSetSw)
unmeth <- getUnmeth(mSetSw)
Mval <- log2((meth + 100)/(unmeth + 100))
beta <- getBeta(mSetSw)
dim(Mval)

# exclude sex chromosomes
meth <- getMeth(mSetFlt)
unmeth <- getUnmeth(mSetFlt)
Mval_flt <- log2((meth + 100)/(unmeth + 100))
beta_flt <- getBeta(mSetFlt)
dim(Mval_flt)

```
